openwisp_users/api/filters.py

```diff
@@ -0,0 +1,159 @@
+import swapper
+from django.core.exceptions import ValidationError
+from rest_framework.exceptions import NotFound
+
+Organization = swapper.load_model('openwisp_users', 'Organization')
+
+
+class FilterByOrganization:
+    """
+    Filter queryset based on the access to the organization
+    of the associated model. Use on of the sub-classes
+    """
+
+    @property
+    def _user_attr(self):
+        raise NotImplementedError()
+
+    def get_queryset(self):
+        qs = super().get_queryset()
+        if self.request.user.is_superuser:
+            return qs
+        return self.get_organization_queryset(qs)
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=getattr(self.request.user, self._user_attr))
+
+
+class FilterByOrganizationMembership(FilterByOrganization):
+    """
+    Filter queryset by organizations the user is a member of
+    """
+
+    _user_attr = 'organizations_dict'
+
+
+class FilterByOrganizationManaged(FilterByOrganization):
+    """
+    Filter queryset by organizations managed by user
+    """
+
+    _user_attr = 'organizations_managed'
+
+
+class FilterByOrganizationOwned(FilterByOrganization):
+    """
+    Filter queryset by organizations owned by user
+    """
+
+    _user_attr = 'organizations_owned'
+
+
+class FilterByParent:
+    """
+    Filter queryset based on one of the parent objects
+    """
+
+    @property
+    def _user_attr(self):
+        raise NotImplementedError()
+
+    def get_queryset(self):
+        qs = super().get_queryset()
+        self.assert_parent_exists()
+        return qs
+
+    def assert_parent_exists(self):
+        parent_queryset = self.get_parent_queryset()
+        if not self.request.user.is_superuser:
+            parent_queryset = self.get_organization_queryset(parent_queryset)
+        try:
+            assert parent_queryset.exists()
+        except (AssertionError, ValidationError):
+            raise NotFound(detail='No relevant data found.')
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=getattr(self.request.user, self._user_attr))
```

Comment:
I don't think we'll need `organization_lookup` here because usually parent objects have the `organization` field defined on them, so this looks good :+1: